File: PROBLEMS/3_Set Matrix Zeroes/Better/soln2.py
# Python Code

def set_matrix_zero(matrix):
    rows = len(matrix)
    cols = len(matrix[0])
    
    row = [0] * rows
    col = [0] * cols

    # First pass: mark rows and columns
    for i in range(rows):
        for j in range(cols):
            if matrix[i][j] == 0:
                row[i] = 1
                col[j] = 1

    # Second pass: set to 0 if marked
    for i in range(rows):
        for j in range(cols):
            if row[i] == 1 or col[j] == 1:
                matrix[i][j] = 0


# Input
rows = int(input("Enter number of rows: "))
cols = int(input("Enter number of columns: "))

matrix = []
print("Enter matrix elements row-wise:")
for _ in range(rows):
    matrix.append(list(map(int, input().split())))

print("\nOriginal Matrix:")
for row in matrix:
    print(" ".join(map(str, row)))

set_matrix_zero(matrix)

print("\nMatrix after setting zeroes:")
for row in matrix:
    print(" ".join(map(str, row)))

File: PROBLEMS/3_Set Matrix Zeroes/Better/soln2.cpp
// C++ Code

#include <iostream>
using namespace std;

void setMatrixZero(int matrix[][100], int rows, int cols) {
    int row[100] = {0};
    int col[100] = {0};

    // Mark rows and columns that contain 0
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (matrix[i][j] == 0) {
                row[i] = 1;
                col[j] = 1;
            }
        }
    }

    // Set zeroes based on row[] and col[]
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (row[i] == 1 || col[j] == 1) {
                matrix[i][j] = 0;
            }
        }
    }
}

int main() {
    int matrix[100][100];
    int rows, cols;

    cout << "Enter rows and columns: ";
    cin >> rows >> cols;

    cout << "Enter matrix elements:\n";
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            cin >> matrix[i][j];
        }
    }

    setMatrixZero(matrix, rows, cols);

    cout << "\nMatrix after setting zeroes:\n";
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            cout << matrix[i][j] << " ";
        }
        cout << endl;
    }

    return 0;
}
